### PR TITLE
Simulated Tests and Validation Functions for Stop Play

### LINKED ISSUE #1908: Implement Simulated Stop Play Test
### Description of the task
Write validation functions to ensure that the robots obey the [rules for stopping](https://robocup-ssl.github.io/ssl-rules/sslrules.html#_stop). Add tests for robots close to ball/far from ball + friendlyhalf/enemyhalf 

### Acceptance criteria

- [ ] write validation functions to check rules
- [ ] identify any broken rules and potential solutions

### PR DESCRIPTION
**Pull request by tarakong01**
### Description

Wrote validation functions and simulated tests for stop play, according to SSL Rules for stopping. Currently commented out all speed checks in tests, linked to #1882 to implement robots slowing down. Last test case is disabled, linked to #1948 to fix bug. Linked other test cases to #1948 as well, for warning messages and bad robot positioning.

Note: will wait on #1946 to merge, before merging this PR (in order to resolve the merge conflicts that will occur due to directory and build file changes from #1946 )

### Testing Done

Simulated tests for the ball at different locations, robots spread out and close together, robots far from the ball and close to the ball. Testing distance from ball and robot speed (commented out for now, see above)

### Resolved Issues

resolves #1908 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/play/stop_play_test.cpp [hunk 1]
```diff
@@ -12,11 +14,12 @@ class StopPlayTest : public SimulatedPlayTestFixture
 {
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/play/stop_play_test.cpp:14]: You could pull some of the common setup functions into the SetUp of this fixture

```
 void SetUp() override
 {
    setFriendlyGoalie(0);
    addEnemyRobots(TestUtil::createStationaryRobotStatesWithId(
        {Point(1, 0), Point(1, 2.5), Point(1, -2.5), field().enemyGoalCenter(),
         field().enemyDefenseArea().negXNegYCorner(),
         field().enemyDefenseArea().negXPosYCorner()}));
    setEnemyGoalie(0);
    setAIPlay(TYPENAME(StopPlay));
    setRefereeCommand(RefereeCommand::STOP, RefereeCommand::STOP);
  }
```

This function will get called before every test

**Review by akhilveeraghanta — COMMENTED** [review]
Nice work on these tests! Glad we are able to catch some of these edge cases

Just a couple comments

File: src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        auto robot_avoids_ball = [robot_id](std::shared_ptr<World> world_ptr) {
+            std::optional<Robot> robotOptional =
+                world_ptr->friendlyTeam().getRobotById(robot_id);
+            if (!robotOptional.has_value())
+            {
+                LOG(FATAL) << "There is no robot with ID: " + std::to_string(robot_id);
+            }
+
+            Robot robot = robotOptional.value();
```
[the remark this answers is not in the document]
Comment (tarakong01) [reply src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.cpp]: Fixed

File: src/software/ai/hl/stp/play/stop_play_test.cpp (hunk 1 above)
Comment (tarakong01) [reply src/software/ai/hl/stp/play/stop_play_test.cpp:14]: Good idea! Fixed

**Review by nimazareian — COMMENTED** [review]
Looks great! I just have a few small comments.

File: src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+                                "m away from the ball!";
+        }
+
+        bool robot_avoided_ball = distance_from_ball >= MIN_DISTANCE_FROM_BALL;
+        if (!robot_avoided_ball)
+        {
+            // TODO: change from throwing exception to FAIL(), once PR #1946 is merged
+            throw std::runtime_error("Robot " + std::to_string(robot_id) +
+                                     " did not avoid the ball!");
+        }
```
[the remark this answers is not in the document]
Comment (tarakong01) [reply src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.cpp]: The first if statement was for debugging, it just printed out the current robot distance/speed as a warning message if it was too close/fast, I forgot to remove it, thanks!

File: src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.h [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+#include "software/simulated_tests/validation/validation_function.h"
+#include "software/world/world.h"
+
+/**
+ * Checks if robots are keeping appropriate distance from
+ * the ball when the Referee has issued a STOP command
+ * @param world_ptr the world pointer given by the simulator. Gets updated every tick
+ * @param yield yields control to the next routine (coroutines)
+ */
```
[the remark this answers is not in the document]
Comment (tarakong01) [reply src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.h]: Good catch, will fix that!

[the remark this answers is not in the document]
Comment (tarakong01) [reply src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.h]: Fixed

File: src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (tarakong01) [reply src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.cpp]: Fixed

File: src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.h [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+#include "software/simulated_tests/validation/validation_function.h"
+#include "software/world/world.h"
+
+/**
+ * Checks if robots are keeping at least 0.5 meters from
+ * the ball
+ * @param world_ptr the world pointer given by the simulator. Gets updated every tick
+ * @param yield yields control to the next routine (coroutines)
+ */
+void robotsAvoidBall(std::shared_ptr<World> world_ptr,
```
Comment (itsarune) [inline src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.h]: can we add a parameter called tolerance or something so that we're not only limited to 0.5m?

**Review by itsarune — COMMENTED** [review]
great tests!

**Review by jonathanlew — COMMENTED** [review]
overall good set of tests, will re-review once #1946 is merged in

File: src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.h (hunk 5 above)
Comment (tarakong01) [reply src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.h]: Can you clarify what you mean by adding tolerance? The stop play requires specifically 0.5m distance, but I could change it so the validation function takes in a parameter rather than using a constant defined in the validation function so that other plays than the stop plays could use this validation function for different distances.

Comment (itsarune) [reply src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.h]: @tarakong01 yup that's great, that's what I meant

File: src/software/simulated_tests/non_terminating_validation_functions/robots_avoid_ball_validation.cpp [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+{
+    for (auto robot : world_ptr->friendlyTeam().getAllRobots())
+    {
+        double distance = (robot.position() - world_ptr->ball().position()).length();
+        const double MIN_DISTANCE_FROM_BALL = ROBOT_MAX_RADIUS_METERS + 0.5;
+
+        if (distance < MIN_DISTANCE_FROM_BALL)
+        {
+            FAIL() << "Robot " + std::to_string(robot.id()) +
+                          " less than 0.5 m away from the ball!";
```
Comment (nimazareian) [inline src/software/simulated_tests/non_terminating_validation_functions/robots_avoid_ball_validation.cpp]: nit: I think it should be "Robot x **_is_** less than ...". Similarly for `robotsSlowDown`.
Other than that it looks great!

File: src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.h (hunk 5 above)
Comment (tarakong01) [reply src/software/simulated_tests/validation_functions/robots_avoid_ball_validation.h]: Added!

File: src/software/simulated_tests/non_terminating_validation_functions/robots_avoid_ball_validation.cpp (hunk 6 above)
Comment (tarakong01) [reply src/software/simulated_tests/non_terminating_validation_functions/robots_avoid_ball_validation.cpp]: Fixed

**Review by nimazareian — DISMISSED** [review]
LGTM!

File: src/software/simulated_tests/non_terminating_validation_functions/robots_slow_down_validation.cpp [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+                    ValidationCoroutine::push_type& yield)
+{
+    for (auto robot : world_ptr->friendlyTeam().getAllRobots())
+    {
+        double speed = robot.velocity().length();
+
+        if (speed > max_speed)
+        {
+            FAIL() << "Robot " + std::to_string(robot.id()) +
+                          " is moving faster than 1.5 m/s!";
```
Comment (akhilveeraghanta) [inline src/software/simulated_tests/non_terminating_validation_functions/robots_slow_down_validation.cpp]: we should print `max_speed` here instead of 1.5

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
Overall LGTM, nice work! just 1 nit

File: src/software/ai/hl/stp/play/stop_play_test.cpp [hunk 8]
```diff
@@ -10,36 +12,225 @@
...
+        [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
+            // Wait 2 seconds for robots that start too close to the ball to move away
+            while (world_ptr->getMostRecentTimestamp() < Timestamp::fromSeconds(2))
+            {
+                yield();
+            }
+            // TODO: (#1882) implement robots slow down when responding to stop command
+            // robotsSlowDown(1.5, world_ptr, yield);
+            robotsAvoidBall(0.5, world_ptr, yield);
+        }};
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/stop_play_test.cpp]: We can make this a class variable in StopPlayTest since we seem to reuse it a lot. We should also give it a name like stop_play_rules

**Review by jonathanlew — CHANGES_REQUESTED** [review]
looks good, just one comment

File: src/software/simulated_tests/non_terminating_validation_functions/robots_slow_down_validation.cpp (hunk 7 above)
Comment (tarakong01) [reply src/software/simulated_tests/non_terminating_validation_functions/robots_slow_down_validation.cpp]: Forgot to change that, thanks for catching it!

Comment (tarakong01) [reply src/software/simulated_tests/non_terminating_validation_functions/robots_slow_down_validation.cpp]: Fixed

File: src/software/ai/hl/stp/play/stop_play_test.cpp (hunk 8 above)
Comment (tarakong01) [reply src/software/ai/hl/stp/play/stop_play_test.cpp]: Fixed

**Review by jonathanlew — APPROVED** [review]
LGTM! :+1:

**Review by akhilveeraghanta — APPROVED** [review]
LGTM :fire:
